File: demo/mob_minion/src/execute_mob_actions.js
/**
 * Execute Mob Actions Module
 * 
 * Implements the Execute Mob Actions sub-epic functionality for commanding mobs
 * and executing coordinated actions for all minions in a mob.
 */

/**
 * MobCommandHandler - Handles identification of mob from clicked token
 * 
 * Story: Click Mob Token To Command
 */
export class MobCommandHandler {
    /**
     * @param {Object} foundryTokenApi - Foundry VTT Token API dependency
     * @param {Object} foundryActorSystem - Foundry VTT Actor System dependency
     */
    constructor(foundryTokenApi, foundryActorSystem) {
        this.foundryTokenApi = foundryTokenApi;
        this.foundryActorSystem = foundryActorSystem;
    }

    /**
     * Identify mob associated with clicked token
     * 
     * Scenario: Game Master clicks mob token to command mob
     * 
     * @param {string} tokenId - ID of clicked token
     * @returns {Object|null} Mob entity if token belongs to mob, null otherwise
     */
    identifyMobFromToken(tokenId) {
        const token = this.foundryTokenApi.getTokenById(tokenId);
        
        if (!token) {
            return null;
        }

        const mobId = token.getFlag('mob-minion', 'mobId');
        
        if (!mobId) {
            return null; // Token doesn't belong to any mob
        }

        return this._getMobEntity(mobId);
    }

    /**
     * Prepare to execute action for all minions in mob
     * 
     * @param {Object} mobEntity - Mob entity to prepare
     * @returns {Object} Action preparation result with mob and minion tokens
     */
    prepareAction(mobEntity) {
        const minionTokens = mobEntity.tokenIds.map(tokenId => 
            this.foundryTokenApi.getTokenById(tokenId)
        ).filter(token => token !== null);

        return {
            mob: mobEntity,
            minionTokens: minionTokens,
            minionCount: minionTokens.length
        };
    }

    /**
     * Get mob entity from Foundry actor system
     * 
     * @private
     * @param {string} mobId - Mob entity ID
     * @returns {Object|null} Mob entity or null if not found
     */
    _getMobEntity(mobId) {
        const actors = this.foundryActorSystem.actors.filter(actor => 
            actor.getFlag('mob-minion', 'mobId') === mobId
        );

        if (actors.length === 0) {
            return null;
        }

        const actor = actors[0];
        return {
            id: actor.getFlag('mob-minion', 'mobId'),
            tokenIds: actor.getFlag('mob-minion', 'tokenIds') || [],
            actorIds: actor.getFlag('mob-minion', 'actorIds') || [],
            strategy: actor.getFlag('mob-minion', 'strategy') || null
        };
    }
}

/**
 * TargetSelector - Determines target based on mob's strategy
 * 
 * Story: Determine Target From Strategy
 */
export class TargetSelector {
    /**
     * @param {Object} foundryCombatSystem - Foundry VTT Combat System dependency
     */
    constructor(foundryCombatSystem) {
        this.foundryCombatSystem = foundryCombatSystem;
    }

    /**
     * Determine target using mob's assigned strategy
     * 
     * Scenario: System determines target using assigned strategy
     * 
     * @param {Object} mobEntity - Mob entity with assigned strategy
     * @returns {Object|null} Selected target enemy or null if none available
     */
    determineTargetFromStrategy(mobEntity) {
        const strategy = mobEntity.strategy || 'Default';
        const availableEnemies = this.foundryCombatSystem.getAvailableEnemies();

        if (availableEnemies.length === 0) {
            return null;
        }

        switch (strategy) {
            case 'AttackMostPowerful':
                return this._selectMostPowerfulTarget(availableEnemies);
            case 'AttackWeakest':
                return this._selectWeakestTarget(availableEnemies);
            case 'AttackMostDamaged':
                return this._selectMostDamagedTarget(availableEnemies);
            case 'DefendLeader':
                return this._selectLeaderTarget(availableEnemies);
            default:
                return this._selectDefaultTarget(availableEnemies);
        }
    }

    /**
     * Select most powerful target (highest power level)
     * 
     * @private
     * @param {Array<Object>} enemies - Available enemy entities
     * @returns {Object} Most powerful enemy
     */
    _selectMostPowerfulTarget(enemies) {
        return enemies.reduce((mostPowerful, enemy) => {
            const enemyPower = this._getEnemyPower(enemy);
            const mostPowerfulPower = this._getEnemyPower(mostPowerful);
            return enemyPower > mostPowerfulPower ? enemy : mostPowerful;
        }, enemies[0]);
    }

    /**
     * Select weakest target (lowest power level)
     * 
     * @private
     * @param {Array<Object>} enemies - Available enemy entities
     * @returns {Object} Weakest enemy
     */
    _selectWeakestTarget(enemies) {
        return enemies.reduce((weakest, enemy) => {
            const enemyPower = this._getEnemyPower(enemy);
            const weakestPower = this._getEnemyPower(weakest);
            return enemyPower < weakestPower ? enemy : weakest;
        }, enemies[0]);
    }

    /**
     * Select most damaged target (highest damage taken)
     * 
     * @private
     * @param {Array<Object>} enemies - Available enemy entities
     * @returns {Object} Most damaged enemy
     */
    _selectMostDamagedTarget(enemies) {
        return enemies.reduce((mostDamaged, enemy) => {
            const enemyDamage = this._getDamageTaken(enemy);
            const mostDamagedDamage = this._getDamageTaken(mostDamaged);
            return enemyDamage > mostDamagedDamage ? enemy : mostDamaged;
        }, enemies[0]);
    }

    /**
     * Select leader target (for defend leader strategy)
     * 
     * @private
     * @param {Array<Object>} enemies - Available enemy entities
     * @returns {Object} Leader enemy (first available)
     */
    _selectLeaderTarget(enemies) {
        // For now, return first enemy as leader
        // This can be enhanced based on specific leader identification logic
        return enemies[0];
    }

    /**
     * Select default target (nearest enemy or first available)
     * 
     * Scenario: System uses default strategy when no strategy assigned
     * 
     * @private
     * @param {Array<Object>} enemies - Available enemy entities
     * @returns {Object} Default target enemy
     */
    _selectDefaultTarget(enemies) {
        // Return first available enemy as default
        return enemies[0];
    }

    /**
     * Get enemy power level
     * 
     * @private
     * @param {Object} enemy - Enemy entity
     * @returns {number} Power level
     */
    _getEnemyPower(enemy) {
        return enemy.system?.attributes?.level?.value || 
               enemy.system?.details?.level || 
               1;
    }

    /**
     * Get damage taken by enemy
     * 
     * @private
     * @param {Object} enemy - Enemy entity
     * @returns {number} Damage taken
     */
    _getDamageTaken(enemy) {
        const maxHealth = enemy.system?.attributes?.hp?.max || 1;
        const currentHealth = enemy.system?.attributes?.hp?.value || maxHealth;
        return maxHealth - currentHealth;
    }
}

/**
 * AttackExecutor - Executes attack actions for all minions in mob
 * 
 * Story: Execute Attack Action
 */
export class AttackExecutor {
    /**
     * @param {Object} foundryCombatSystem - Foundry VTT Combat System dependency
     */
    constructor(foundryCombatSystem) {
        this.foundryCombatSystem = foundryCombatSystem;
    }

    /**
     * Execute attack for all minions in mob via Foundry combat system
     * 
     * Scenario: System executes attack for all minions in mob
     * 
     * @param {Object} actionPreparation - Action preparation result from MobCommandHandler
     * @param {Object} target - Target enemy entity
     * @returns {Array<Object>} Array of attack results for each minion
     */
    executeAttackAction(actionPreparation, target) {
        const { minionTokens } = actionPreparation;
        const attackResults = [];

        minionTokens.forEach(minionToken => {
            const attackResult = this._executeMinionAttack(minionToken, target);
            attackResults.push(attackResult);
        });

        return attackResults;
    }

    /**
     * Execute attack for a single minion
     * 
     * @private
     * @param {Object} minionToken - Minion token entity
     * @param {Object} target - Target enemy entity
     * @returns {Object} Attack result with hit/miss and damage
     */
    _executeMinionAttack(minionToken, target) {
        // Use Foundry combat system to execute attack
        const attackRoll = this.foundryCombatSystem.rollAttack(minionToken, target);
        const damageRoll = attackRoll.hit ? 
            this.foundryCombatSystem.rollDamage(minionToken, target) : 
            null;

        return {
            minionId: minionToken.id,
            targetId: target.id,
            hit: attackRoll.hit,
            damage: damageRoll?.total || 0
        };
    }
}

File: demo/mob_minion/src/execute_mob_actions.py
"""
Execute Mob Actions Module

Implements the Execute Mob Actions sub-epic functionality for commanding mobs
and executing coordinated actions for all minions in a mob.
"""
from typing import List, Dict, Optional
from .create_mob import Mob


class MobCommandHandler:
    """Handles identification of mob from clicked token.
    
    Story: Click Mob Token To Command
    """
    
    def __init__(self, foundry_token_api, foundry_actor_system):
        """Initialize MobCommandHandler with dependencies.
        
        Args:
            foundry_token_api: Foundry VTT Token API dependency
            foundry_actor_system: Foundry VTT Actor System dependency
        """
        self.foundry_token_api = foundry_token_api
        self.foundry_actor_system = foundry_actor_system
    
    def identify_mob_from_token(self, token_id: str) -> Optional[Mob]:
        """Identify mob associated with clicked token.
        
        Scenario: Game Master clicks mob token to command mob
        
        Args:
            token_id: ID of clicked token
            
        Returns:
            Optional[Mob]: Mob entity if token belongs to mob, None otherwise
        """
        token = self.foundry_token_api.get_token_by_id(token_id)
        
        if not token:
            return None
        
        mob_id = token.get_flag('mob-minion', 'mobId')
        
        if not mob_id:
            return None  # Token doesn't belong to any mob
        
        return self._get_mob_entity(mob_id)
    
    def prepare_action(self, mob: Mob) -> Dict:
        """Prepare to execute action for all minions in mob.
        
        Args:
            mob: Mob entity to prepare
            
        Returns:
            Dict: Action preparation result with mob and minion tokens
        """
        minion_tokens = [
            self.foundry_token_api.get_token_by_id(token_id)
            for token_id in mob.token_ids
        ]
        minion_tokens = [token for token in minion_tokens if token is not None]
        
        return {
            'mob': mob,
            'minion_tokens': minion_tokens,
            'minion_count': len(minion_tokens)
        }
    
    def _get_mob_entity(self, mob_id: str) -> Optional[Mob]:
        """Get mob entity from Foundry actor system.
        
        Args:
            mob_id: Mob entity ID
            
        Returns:
            Optional[Mob]: Mob entity or None if not found
        """
        mob_data = self.foundry_actor_system.get_mob(mob_id)
        
        if not mob_data:
            return None
        
        return Mob(
            mob_id=mob_data.id,
            token_ids=mob_data.token_ids,
            actor_ids=mob_data.actor_ids
        )


class TargetSelector:
    """Determines target based on mob's strategy.
    
    Story: Determine Target From Strategy
    """
    
    def __init__(self, foundry_combat_system):
        """Initialize TargetSelector with dependency.
        
        Args:
            foundry_combat_system: Foundry VTT Combat System dependency
        """
        self.foundry_combat_system = foundry_combat_system
    
    def determine_target_from_strategy(self, mob: Mob) -> Optional[Dict]:
        """Determine target using mob's assigned strategy.
        
        Scenario: System determines target using assigned strategy
        
        Args:
            mob: Mob entity with assigned strategy
            
        Returns:
            Optional[Dict]: Selected target enemy or None if none available
        """
        strategy = mob.strategy or 'Default'
        available_enemies = self.foundry_combat_system.get_available_enemies()
        
        if len(available_enemies) == 0:
            return None
        
        if strategy == 'AttackMostPowerful':
            return self._select_most_powerful_target(available_enemies)
        elif strategy == 'AttackWeakest':
            return self._select_weakest_target(available_enemies)
        elif strategy == 'AttackMostDamaged':
            return self._select_most_damaged_target(available_enemies)
        elif strategy == 'DefendLeader':
            return self._select_leader_target(available_enemies)
        else:
            return self._select_default_target(available_enemies)
    
    def _select_most_powerful_target(self, enemies: List[Dict]) -> Dict:
        """Select most powerful target (highest power level).
        
        Args:
            enemies: Available enemy entities
            
        Returns:
            Dict: Most powerful enemy
        """
        return max(enemies, key=lambda e: self._get_enemy_power(e))
    
    def _select_weakest_target(self, enemies: List[Dict]) -> Dict:
        """Select weakest target (lowest power level).
        
        Args:
            enemies: Available enemy entities
            
        Returns:
            Dict: Weakest enemy
        """
        return min(enemies, key=lambda e: self._get_enemy_power(e))
    
    def _select_most_damaged_target(self, enemies: List[Dict]) -> Dict:
        """Select most damaged target (highest damage taken).
        
        Args:
            enemies: Available enemy entities
            
        Returns:
            Dict: Most damaged enemy
        """
        return max(enemies, key=lambda e: self._get_damage_taken(e))
    
    def _select_leader_target(self, enemies: List[Dict]) -> Dict:
        """Select leader target (for defend leader strategy).
        
        Args:
            enemies: Available enemy entities
            
        Returns:
            Dict: Leader enemy (first available)
        """
        return enemies[0]
    
    def _select_default_target(self, enemies: List[Dict]) -> Dict:
        """Select default target (nearest enemy or first available).
        
        Scenario: System uses default strategy when no strategy assigned
        
        Args:
            enemies: Available enemy entities
            
        Returns:
            Dict: Default target enemy
        """
        return enemies[0]
    
    def _get_enemy_power(self, enemy: Dict) -> int:
        """Get enemy power level.
        
        Args:
            enemy: Enemy entity
            
        Returns:
            int: Power level
        """
        return enemy.get('powerLevel', enemy.get('level', 1))
    
    def _get_damage_taken(self, enemy: Dict) -> int:
        """Get damage taken by enemy.
        
        Args:
            enemy: Enemy entity
            
        Returns:
            int: Damage taken
        """
        max_health = enemy.get('maxHealth', 1)
        current_health = enemy.get('currentHealth', max_health)
        return max_health - current_health


class AttackExecutor:
    """Executes attack actions for all minions in mob.
    
    Story: Execute Attack Action
    """
    
    def __init__(self, foundry_combat_system):
        """Initialize AttackExecutor with dependency.
        
        Args:
            foundry_combat_system: Foundry VTT Combat System dependency
        """
        self.foundry_combat_system = foundry_combat_system
    
    def execute_attack_action(self, action_preparation: Dict, target: Dict) -> List[Dict]:
        """Execute attack for all minions in mob via Foundry combat system.
        
        Scenario: System executes attack for all minions in mob
        
        Args:
            action_preparation: Action preparation result from MobCommandHandler
            target: Target enemy entity
            
        Returns:
            List[Dict]: Array of attack results for each minion
        """
        mob = action_preparation['mob']
        minion_tokens = action_preparation['minion_tokens']
        attack_results = []
        
        for minion_token in minion_tokens:
            attack_result = self._execute_minion_attack(minion_token, target)
            attack_results.append(attack_result)
        
        return attack_results
    
    def _execute_minion_attack(self, minion_token: Dict, target: Dict) -> Dict:
        """Execute attack for a single minion.
        
        Args:
            minion_token: Minion token entity
            target: Target enemy entity
            
        Returns:
            Dict: Attack result with hit/miss and damage
        """
        attack_roll = self.foundry_combat_system.roll_attack(minion_token, target)
        damage_roll = None
        
        if attack_roll.get('hit', False):
            damage_roll = self.foundry_combat_system.roll_damage(minion_token, target)
        
        return {
            'minion_id': minion_token.get('id'),
            'target_id': target.get('id'),
            'hit': attack_roll.get('hit', False),
            'damage': damage_roll.get('total', 0) if damage_roll else 0
        }
